src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,298 @@
+
+template<>
+MSHADOW_XINLINE float gesvd_back_helper_eps(float* X) {
+  return 1e-30;
+}
+
+template<>
+MSHADOW_XINLINE double gesvd_back_helper_eps(double* X) {
+  return 1e-100;
+}
+
+struct GesvdBackHelper_dV {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* L, int ldl,
+                                  DType* dA, int ldda) {
+    const int offl(k * ldl);
+    const int offda(k * m * ldda);
+    DType denom(0.0);
+    const DType eps(gesvd_back_helper_eps(dA));
+    for (int i = 0; i < m; ++i) {
+      denom = L[offl + i];
+      if (denom < eps) denom = eps;
+      for (int j = 0; j < n; ++j) {
+        dA[offda + i * ldda + j] /= denom;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G1 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl) {
+    const int offl(k * ldl);
+    const int offx(k * m * ldx);
+    DType numer(0.0);
+    for (int i = 0; i < m; ++i) {
+      for (int j = 0; j < m; ++j) {
+        numer = L[offl + j];
+        X[offx + i * ldx + j] *= numer;
+      }
+    }
+  }
+};
+
+struct GesvdBackHelper_G2 {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int k, int m, int n, DType* X, int ldx,
+                                  DType* L, int ldl, DType* dL, int lddl,
+                                  DType* dA, int ldda, DType* V, int ldv) {
+    const int offx(k * m * ldx);
+    const int offl(k * ldl);
+    const int offdl(k * lddl);
+    const int offda(k * m * ldda);
+    const int offv(k * m * ldv);
+    const DType eps(gesvd_back_helper_eps(X));
+    DType denom1(0.0), denom2(0.0), elem(0.0);
+
+    for (int i = 0; i < m - 1; ++i) {
+      for (int j = i + 1; j < m; ++j) {
+        denom1 = L[offl + i] - L[offl + j];
+        denom2 = L[offl + i] + L[offl + j];
+        if (denom1 < eps) denom1 = eps;
+        if (denom2 < eps) denom2 = eps;
+        elem = (X[offx + i * ldx + j] - X[offx + j * ldx + i]) / denom1 / denom2;
+        X[offx + i * ldx + j] = elem * L[offl + j];
+        X[offx + j * ldx + i] = elem * L[offl + i];
+      }
+    }
+    for (int i = 0; i < m; ++i) {
+      elem = DType(0.0);
+      for (int j = 0; j < n; ++j) {
+        elem += dA[offda + i * ldda + j] * V[offv + i * ldv + j];
+      }
+      elem = -elem + dL[offdl + i];
+      X[offx + i * ldx + i] = elem;
+    }
+  }
+};
+
+struct gesvd_backward {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 3, DType>& dUT,
+                 const Tensor<xpu, 2, DType>& dL,
+                 const Tensor<xpu, 3, DType>& dV,
+                 const Tensor<xpu, 3, DType>& UT,
+                 const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 3, DType>& V,
+                 const Tensor<xpu, 3, DType>& dA,
+                 const Tensor<xpu, 3, DType>& tempMs,
+                 const Tensor<xpu, 3, DType>& tempMr,
```

Comment:
Could you elaborate how to 
"coding up "in-place" dA <- dot(UT, dA) using a temp space of shape (m, m)"?
The method I have thought about is:
Assume n % m == 0 for simplicity.
We can write dA as blocks: dA = [dA1, dA2, ..., dAx], where dAi is of shape (m, m) and x = n / m
so dot(UT, dA) = [dot(UT, dA1), ..., dot(UT, dAx], and each dot can be achieved with temp space of shape (m, m)
I don't know whether I have understood correctly.

Reply:
Yes exactly, something like you are saying. Essentially you slice the large matrix up into (m, m) blocks, and then you can even just call gemm in a loop. This may be the easiest, in fact. What I mean, you mask out (m, m) blocks of dA and iterate over them, always replacing one block B with dot(UT, B). For that, you need an (m, m) temp space, but that you have.

Reply:
Please ask if this is still unclear (but your comment is what I have in mind). You can mask out blocks of dA simply by moving the pointer, everything else (stride, etc) remains the same, because the n-axis is the continuous one. The final block will be (m, m2), where m2 <= m, but that should also not be a problem.

Reply:
Now my implementation loops over the m/n blocks (each of shape (m, m) and reduce the temp space needed to (m, m).